File: python/src/adeu/domain.py
import re
from typing import Any, Dict, List, Tuple

from docx.oxml.ns import qn
from docx.text.paragraph import Paragraph
from rapidfuzz.distance import Levenshtein

from adeu.utils.docx import get_run_text, iter_block_items


def _get_paragraph_text(p: Paragraph) -> str:
    return "".join(get_run_text(r) for r in p.runs)


def levenshtein_distance(s1: str, s2: str) -> int:
    """C-backed Levenshtein via rapidfuzz. ~50-100x faster than pure Python."""
    return Levenshtein.distance(s1, s2)


_TERM_BODY = r"[A-Z][A-Za-z0-9\s\-&\'’]{1,60}"
_LEADING_TERM_RE = re.compile(rf"^(?:[\d\.\-\(\)a-zA-Z]+\s*)?[\"“]({_TERM_BODY})[\"”]")
_SENTENCE_TERM_RE = re.compile(rf"(?<=[\.\;\:\!\?])\s+(?:[\d\.\-\(\)a-zA-Z]+\s+)?[\"“]({_TERM_BODY})[\"”]")
_INLINE_TERM_RE = re.compile(rf'\([^)]*?["“]({_TERM_BODY})["”][^)]*?\)')


def extract_terms_from_paragraph(text: str) -> List[str]:
    """
    All defined terms declared in one paragraph, in appearance order,
    deduplicated. Language-agnostic: keyed on quoting typography, never on
    English phrases like "means".
    """
    found: List[Tuple[int, str]] = []
    leading = _LEADING_TERM_RE.match(text)
    if leading:
        found.append((leading.start(1), leading.group(1).strip()))
    for m in _SENTENCE_TERM_RE.finditer(text):
        found.append((m.start(1), m.group(1).strip()))
    for m in _INLINE_TERM_RE.finditer(text):
        found.append((m.start(1), m.group(1).strip()))

    terms: List[str] = []
    seen_positions: set = set()
    for pos, term in sorted(found):
        if pos in seen_positions:
            continue
        seen_positions.add(pos)
        terms.append(term)
    return terms


# FILE: src/adeu/domain.py
def extract_document_settings_warnings(doc) -> List[str]:
    """
    Inspects word/settings.xml for privacy flags that cause Microsoft Word to
    silently strip attribution from tracked changes and comments the next time
    the document is opened and saved.

    The engine itself preserves the attribution it writes — but Word will
    destroy `w:author`, `w:initials`, and/or `w:date` on next save when these
    flags are enabled. We surface this to the agent at read time so it can
    decide how to handle the situation before making edits.

    Returns a list of warning strings (one per enabled flag), without leading
    bullets. Empty list if settings.xml is absent or neither flag is enabled.

    OOXML boolean truthiness:
      - element absent           → disabled
      - element present, no w:val → enabled (OOXML default)
      - element with w:val in {"0","false","off"} (case-insensitive) → disabled
      - element with any other w:val → enabled
    """
    warnings: List[str] = []

    settings_part = None
    for part in doc.part.package.parts:
        if str(part.partname) == "/word/settings.xml":
            settings_part = part
            break
    if settings_part is None:
        return warnings

    # The settings part may be a generic Part or an XmlPart depending on how
    # python-docx loaded it. Parse the blob directly so this is robust either way.
    from docx.oxml import parse_xml

    try:
        root = parse_xml(settings_part.blob)
    except Exception:
        return warnings

    def _is_truthy(el) -> bool:
        # OOXML boolean rule: element present with no w:val attribute defaults to true.
        # w:val of "0", "false", or "off" (case-insensitive) means disabled.
        val = el.get(qn("w:val"))
        if val is None:
            return True
        return val.lower() not in ("0", "false", "off")

    # Use iter() + local name matching for namespace-agnostic lookup, mirroring
    # the TS findDescendantsByLocalName helper. This is robust to settings.xml
    # variants from different Word versions even though w:settings is the
    # canonical namespace.
    def _find_first_by_local_name(local_name: str):
        suffix = "}" + local_name
        for el in root.iter():
            tag = el.tag
            if isinstance(tag, str) and (tag == local_name or tag.endswith(suffix)):
                return el
        return None

    remove_personal = _find_first_by_local_name("removePersonalInformation")
    if remove_personal is not None and _is_truthy(remove_personal):
        warnings.append(
            "[Warning] Privacy flag `removePersonalInformation` is enabled in word/settings.xml. "
            "Microsoft Word will strip the `w:author`, `w:initials`, and `w:date` attributes "
            "from every tracked change and comment the next time this document is opened and saved. "
            "Edits made by this agent will lose attribution, breaking audit trails and any "
            "multi-turn workflow that relies on identifying prior edits."
        )

    remove_date_time = _find_first_by_local_name("removeDateAndTime")
    if remove_date_time is not None and _is_truthy(remove_date_time):
        warnings.append(
            "[Warning] Privacy flag `removeDateAndTime` is enabled in word/settings.xml. "
            "Microsoft Word will strip the `w:date` attribute from every tracked change and "
            "comment the next time this document is opened and saved. Timestamps on this "
            "agent's edits will be lost on the next Word save."
        )

    return warnings


def extract_all_domain_metadata(
    doc, base_text: str
) -> Tuple[Dict[str, Dict[str, Any]], List[str], Dict[str, Dict[str, Any]]]:
    """
    SINGLE-PASS EXTRACTION ENGINE
    Replaces multiple document walks by simultaneously extracting Definitions,
    Bookmarks, and Cross-References in a single O(N) iteration over block items.
    """
    definitions: Dict[str, Dict[str, Any]] = {}
    duplicates = set()
    raw_anchors: Dict[str, Dict[str, Any]] = {}
    raw_references: List[Tuple[str, str]] = []  # (target_bookmark, referencing_text)

    # --- SINGLE DOCUMENT WALK ---
    for item in iter_block_items(doc):
        if not isinstance(item, Paragraph):
            continue

        # Used for definition extraction and anchor previews
        text = _get_paragraph_text(item).strip()
        if not text:
            continue

        # 1. Extract Definitions (every declaration in the paragraph — QA M7)
        for term in extract_terms_from_paragraph(text):
            if term in definitions:
                duplicates.add(term)
            else:
                definitions[term] = {"count": 0}

        # 2. Extract Bookmarks & References simultaneously
        short_text = text[:60] + ("..." if len(text) > 60 else "")

        for node in item._element.iter():
            # Check Bookmarks
            if node.tag == qn("w:bookmarkStart"):
                b_name = node.get(qn("w:name"))
                if b_name and (not b_name.startswith("_") or b_name.startswith("_Ref")):
                    if b_name not in raw_anchors:
                        raw_anchors[b_name] = {
                            "anchored_to": short_text,
                            "referenced_from": [],
                        }

            # Check References
            target = None
            if node.tag == qn("w:fldSimple"):
                instr = node.get(qn("w:instr"), "")
                parts = instr.strip().split()
                if parts and parts[0] == "REF" and len(parts) > 1:
                    target = parts[1]
            elif node.tag == qn("w:instrText"):
                instr = node.text or ""
                parts = instr.strip().split()
                if parts and parts[0] == "REF" and len(parts) > 1:
                    target = parts[1]

            if target:
                raw_references.append((target, short_text))

    # --- POST-PROCESSING: Resolve Relationships ---
    for target, ref_text in raw_references:
        if target in raw_anchors:
            raw_anchors[target]["referenced_from"].append(ref_text)

    # --- POST-PROCESSING: Definitions & Diagnostics ---
    diagnostics = []

    if definitions:
        sorted_terms = sorted(definitions.keys(), key=len, reverse=True)
        alt = "|".join(re.escape(t) for t in sorted_terms)
        usage_pattern = re.compile(rf'(?<!["“])\b({alt})\b(?!["”])')

        for m in usage_pattern.finditer(base_text):
            matched_term = m.group(1)
            if matched_term in definitions:
                definitions[matched_term]["count"] += 1

        # Symbol-table noise reduction must not gate diagnostics — see the
        # matching comment in extract_definitions_and_diagnostics (QA F6).
        for term in list(definitions.keys()):
            if definitions[term]["count"] == 0:
                del definitions[term]
                if term not in duplicates:
                    diagnostics.append(f"[Warning] Unused Definition: '{term}' is defined but never used.")

    for term in duplicates:
        diagnostics.append(f"[Error] Duplicate Definition: '{term}' is defined multiple times.")

    stop_words = {
        "The",
        "This",
        "That",
        "Such",
        "A",
        "An",
        "Any",
        "All",
        "Some",
        "No",
        "Every",
        "Each",
        "As",
        "In",
        "Of",
        "For",
        "To",
        "On",
        "By",
        "With",
    }

    all_cap_pattern = r"\b[A-Z][a-zA-Z]*(?:\s+[A-Z][a-zA-Z]*)*\b"
    all_caps = set(re.findall(all_cap_pattern, base_text))

    valid_terms = set(definitions.keys())
    terms_by_first_letter: Dict[str, List[str]] = {}
    for term in valid_terms:
        terms_by_first_letter.setdefault(term[0].lower(), []).append(term)

    candidates_by_term: Dict[str, List[str]] = {}

    for candidate in all_caps:
        candidate = candidate.strip()
        words = candidate.split()
        while words and words[0].title() in stop_words:
            words = words[1:]
        candidate = " ".join(words)

        if len(candidate) < 4:
            continue
        if candidate in valid_terms:
            continue

        first_letter = candidate[0].lower()
        candidate_terms = terms_by_first_letter.get(first_letter, [])

        if len(candidate) > 5:
            for k, v in terms_by_first_letter.items():
                if k != first_letter:
                    candidate_terms = candidate_terms + v

        for term in candidate_terms:
            if abs(len(candidate) - len(term)) > 2:
                continue
            if candidate == term + "s" or candidate == term + "es":
                continue
            if term == candidate + "s" or term == candidate + "es":
                continue

            dist = Levenshtein.distance(candidate, term, score_cutoff=2)
            if dist == 0 or dist > 2:
                continue

            if len(term) <= 5:
                if dist > 1:
                    continue
                if candidate[0].lower() != term[0].lower():
                    continue

            if term not in candidates_by_term:
                candidates_by_term[term] = []
            if candidate not in candidates_by_term[term]:
                candidates_by_term[term].append(candidate)

    for term, candidates in candidates_by_term.items():
        c_str = ", ".join(f"'{c}'" for c in sorted(candidates))
        diagnostics.append(f"[Info] Possible Typos for '{term}': Found {c_str}")

    def diag_sort_key(msg):
        if msg.startswith("[Error]"):
            return 0
        if msg.startswith("[Warning]"):
            return 1
        return 2

    diagnostics.sort(key=lambda x: (diag_sort_key(x), x))

    return definitions, diagnostics, raw_anchors


def _content_controls_appendix_section(doc, base_text: str) -> List[str]:
    """The appendix's ``## Content Controls`` block, or [] when unwarranted.

    Imported lazily: ``adeu.fields`` pulls in the ordinal pre-pass, and the
    appendix is built during ingest for every read, including documents with no
    controls at all.
    """
    try:
        from adeu.fields import (
            field_summary,
            read_document_protection,
            render_appendix_section,
        )

        # Counts only. The appendix renders four numbers, and computing the
        # full ledger to get them added 115ms of unrendered value previews and
        # breadcrumbs to every read of a control-heavy document.
        counts = field_summary(doc)
        protection = read_document_protection(doc)
    except Exception:
        # The appendix is advisory. A malformed settings part or an exotic
        # control must not take down every read of the document.
        return []
    return render_appendix_section(
        counts,
        protection,
        hint='Read with mode="fields" for the full field ledger.',
    )


def build_structural_appendix(doc, base_text: str) -> str:
    """
    Compiles the Read-Only Structural Appendix block for the agent.
    Returns an empty string if no relevant domain metadata is found.
    """
    defs, diagnostics, anchors = extract_all_domain_metadata(doc, base_text)
    settings_warnings = extract_document_settings_warnings(doc)
    lines: List[str] = [
        "\n\n---",
        "",
        "<!-- READONLY_BOUNDARY_START -->",
        "# Document Structure (Read-Only)",
        (
            "The content below is metadata describing the document's reference structure. "
            "Do not include this section in any tracked changes or edits — it is for your "
            "context only and will be discarded on write."
        ),
    ]

    has_content = False

    if settings_warnings:
        has_content = True
        lines.append("\n## Document Settings")
        for warning in settings_warnings:
            lines.append(f"- {warning}")

    # Spec-fields-ledger §5: the HEADER LINES ONLY. The full ledger never
    # renders here — FedRAMP rev4 has 5,007 controls and the appendix is
    # bounded, so a ledger would swallow every other section.
    cc_lines = _content_controls_appendix_section(doc, base_text)
    if cc_lines:
        has_content = True
        lines.append("")
        lines.extend(cc_lines)

    if defs:
        has_content = True
        lines.append("\n## Defined Terms")
        for term, data in defs.items():
            count = data["count"]
            lines.append(f'- "{term}" — used {count} time{"" if count == 1 else "s"}.')

    if diagnostics:
        has_content = True
        lines.append("\n## Semantic Diagnostics")
        for diag in diagnostics:
            lines.append(f"- {diag}")

    if anchors:
        has_content = True
        lines.append("\n## Named Anchors")
        for b_name, data in anchors.items():
            lines.append(f'- {b_name} → Anchored to: "{data["anchored_to"]}"')
            for ref in data["referenced_from"]:
                lines.append(f'  - Referenced from: "{ref}"')

    if has_content:
        return "\n".join(lines)
    return ""

File: node/packages/core/src/domain.ts
import { DocumentObject } from "./docx/bridge.js";
import { Paragraph, Run } from "./docx/primitives.js";
import { iter_block_items, get_run_text } from "./utils/docx.js";
import { findAllDescendants } from "./docx/dom.js";
import { findDescendantsByLocalName } from "./sanitize/transforms.js";
import {
  fieldSummary,
  readDocumentProtection,
  renderAppendixSection,
} from "./fields.js";

function boundedLevenshtein(a: string, b: string, maxDist: number = 2): number {
  if (a === b) return 0;
  if (Math.abs(a.length - b.length) > maxDist) return maxDist + 1;
  if (a.length === 0) return b.length <= maxDist ? b.length : maxDist + 1;
  if (b.length === 0) return a.length <= maxDist ? a.length : maxDist + 1;

  if (a.length > b.length) {
    const temp = a;
    a = b;
    b = temp;
  }

  let row = Array.from({ length: a.length + 1 }, (_, i) => i);

  for (let i = 1; i <= b.length; i++) {
    const newRow = [i];
    let minInRow = i;
    for (let j = 1; j <= a.length; j++) {
      const cost = a[j - 1] === b[i - 1] ? 0 : 1;
      const val = Math.min(row[j] + 1, newRow[j - 1] + 1, row[j - 1] + cost);
      newRow.push(val);
      if (val < minInRow) minInRow = val;
    }
    if (minInRow > maxDist) return maxDist + 1;
    row = newRow;
  }
  return row[a.length] <= maxDist ? row[a.length] : maxDist + 1;
}

function _get_paragraph_text(p: Paragraph): string {
  let text = "";
  const runs = findAllDescendants(p._element, "w:r");
  for (const r of runs) {
    text += get_run_text(new Run(r, p));
  }
  return text;
}

const _TERM_BODY = "[A-Z][A-Za-z0-9\\s\\-&'’]{1,60}";
// Definition typography, matched repeatedly within a paragraph (QA 2026-07-18
// M7 — a paragraph defining Alpha, Beta AND Gamma must yield all three):
//   1. paragraph-leading quoted term (optionally after a numbering token)
//   2. sentence-leading quoted term (after . ; : ! ?)
//   3. parenthesized inline definition — (the "Term")
// The 'd' flag (ES2022 match indices) recovers each term's exact position for
// position-keyed dedupe, mirroring Python's m.start(1).
const _LEADING_TERM_RE = new RegExp(
  `^(?:[\\d.\\-()a-zA-Z]+\\s*)?["“](${_TERM_BODY})["”]`,
  "d",
);
// Like the leading pattern, a sentence-start definition may carry a numbering
// token ('… the product. 2.2 "Beta" means …').
const _SENTENCE_TERM_RE = new RegExp(
  `(?<=[.;:!?])\\s+(?:[\\d.\\-()a-zA-Z]+\\s+)?["“](${_TERM_BODY})["”]`,
  "dg",
);
const _INLINE_TERM_RE = new RegExp(
  `\\([^)]*?["“](${_TERM_BODY})["”][^)]*?\\)`,
  "dg",
);

/**
 * All defined terms declared in one paragraph, in appearance order,
 * deduplicated. Language-agnostic: keyed on quoting typography, never on
 * English phrases like "means".
 */
export function extract_terms_from_paragraph(text: string): string[] {
  const found: [number, string][] = [];
  const leading = _LEADING_TERM_RE.exec(text) as RegExpExecArray & {
    indices?: Array<[number, number]>;
  };
  if (leading && leading.indices && leading.indices[1]) {
    found.push([leading.indices[1][0], leading[1].trim()]);
  }
  for (const m of text.matchAll(_SENTENCE_TERM_RE)) {
    const indices = (m as any).indices as Array<[number, number]> | undefined;
    if (indices && indices[1]) found.push([indices[1][0], m[1].trim()]);
  }
  for (const m of text.matchAll(_INLINE_TERM_RE)) {
    const indices = (m as any).indices as Array<[number, number]> | undefined;
    if (indices && indices[1]) found.push([indices[1][0], m[1].trim()]);
  }

  const terms: string[] = [];
  const seen_positions = new Set<number>();
  found.sort((a, b) => a[0] - b[0]);
  for (const [pos, term] of found) {
    if (seen_positions.has(pos)) continue;
    seen_positions.add(pos);
    terms.push(term);
  }
  return terms;
}

export function extract_all_domain_metadata(
  doc: DocumentObject,
  base_text: string,
): [
  Record<string, { count: number }>,
  string[],
  Record<string, { anchored_to: string; referenced_from: string[] }>,
] {
  // Null-prototype throughout: these dictionaries are keyed on strings taken
  // from the document (defined terms, w:bookmarkStart/@w:name), so a name like
  // "toString" or "constructor" must be an ordinary entry. On a `{}` literal
  // the presence checks below read Object.prototype instead: the anchor is
  // never registered, and the REF back-fill at the bottom of this function then
  // dies in `raw_anchors[target].referenced_from.push(...)`.
  const definitions: Record<string, { count: number }> = Object.create(null);
  const duplicates = new Set<string>();
  const raw_anchors: Record<
    string,
    { anchored_to: string; referenced_from: string[] }
  > = Object.create(null);
  const raw_references: [string, string][] = [];

  for (const item of iter_block_items(doc)) {
    if (!(item instanceof Paragraph)) continue;

    const text = _get_paragraph_text(item).trim();
    if (!text) continue;

    // 1. Extract Definitions (every declaration in the paragraph — QA M7)
    for (const term of extract_terms_from_paragraph(text)) {
      if (definitions[term]) duplicates.add(term);
      else definitions[term] = { count: 0 };
    }

    const short_text = text.length > 60 ? text.substring(0, 60) + "..." : text;

    const nodes = findAllDescendants(item._element, "*");
    for (const node of nodes) {
      if (node.tagName === "w:bookmarkStart") {
        const b_name = node.getAttribute("w:name");
        if (b_name && (!b_name.startsWith("_") || b_name.startsWith("_Ref"))) {
          if (!raw_anchors[b_name]) {
            raw_anchors[b_name] = {
              anchored_to: short_text,
              referenced_from: [],
            };
          }
        }
      }

      let target: string | null = null;
      if (node.tagName === "w:fldSimple") {
        const instr = node.getAttribute("w:instr") || "";
        const parts = instr.trim().split(/\s+/);
        if (parts.length > 1 && parts[0] === "REF") target = parts[1];
      } else if (node.tagName === "w:instrText") {
        const instr = node.textContent || "";
        const parts = instr.trim().split(/\s+/);
        if (parts.length > 1 && parts[0] === "REF") target = parts[1];
      }

      if (target) raw_references.push([target, short_text]);
    }
  }

  for (const [target, ref_text] of raw_references) {
    if (raw_anchors[target]) {
      raw_anchors[target].referenced_from.push(ref_text);
    }
  }

  const diagnostics: string[] = [];

  const def_keys = Object.keys(definitions);
  if (def_keys.length > 0) {
    const sorted_terms = def_keys.sort((a, b) => b.length - a.length);
    const escapeRegExp = (str: string) =>
      str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const alt = sorted_terms.map(escapeRegExp).join("|");
    const usage_pattern = new RegExp(`(?<!["“])\\b(${alt})\\b(?![”"])`, "g");

    for (const m of base_text.matchAll(usage_pattern)) {
      const matched_term = m[1];
      if (definitions[matched_term]) definitions[matched_term].count++;
    }

    // Drop unused terms from the SYMBOL TABLE only — the filter is noise
    // reduction for the Defined Terms listing, and must not gate the
    // Semantic Diagnostics: a term defined twice and never used is two
    // drafting errors, not zero (QA 2026-07-17 F6; mirrors Python). Surface
    // the orphan definition itself as a diagnostic instead.
    for (const term of def_keys) {
      if (definitions[term].count === 0) {
        delete definitions[term];
        if (!duplicates.has(term)) {
          diagnostics.push(
            `[Warning] Unused Definition: '${term}' is defined but never used.`,
          );
        }
      }
    }
  }

  for (const term of duplicates) {
    diagnostics.push(
      `[Error] Duplicate Definition: '${term}' is defined multiple times.`,
    );
  }

  const stop_words = new Set([
    "The",
    "This",
    "That",
    "Such",
    "A",
    "An",
    "Any",
    "All",
    "Some",
    "No",
    "Every",
    "Each",
    "As",
    "In",
    "Of",
    "For",
    "To",
    "On",
    "By",
    "With",
  ]);

  const all_cap_pattern = /\b[A-Z][a-zA-Z]*(?:\s+[A-Z][a-zA-Z]*)*\b/g;
  const all_caps = new Set(base_text.match(all_cap_pattern) || []);

  const valid_terms = new Set(Object.keys(definitions));
  const terms_by_first_letter: Record<string, string[]> = {};
  for (const term of valid_terms) {
    const fl = term[0].toLowerCase();
    if (!terms_by_first_letter[fl]) terms_by_first_letter[fl] = [];
    terms_by_first_letter[fl].push(term);
  }

  const candidates_by_term: Record<string, string[]> = Object.create(null);

  for (const raw_candidate of all_caps) {
    let candidate = raw_candidate.trim();
    const words = candidate.split(/\s+/);
    while (words.length > 0) {
      const first = words[0];
      const title =
        first.charAt(0).toUpperCase() + first.slice(1).toLowerCase();
      if (stop_words.has(title)) words.shift();
      else break;
    }
    candidate = words.join(" ");

    if (candidate.length < 4) continue;
    if (valid_terms.has(candidate)) continue;

    const first_letter = candidate[0].toLowerCase();
    let candidate_terms = terms_by_first_letter[first_letter] || [];

    if (candidate.length > 5) {
      for (const [k, v] of Object.entries(terms_by_first_letter)) {
        if (k !== first_letter) candidate_terms = candidate_terms.concat(v);
      }
    }

    for (const term of candidate_terms) {
      if (Math.abs(candidate.length - term.length) > 2) continue;
      if (candidate === term + "s" || candidate === term + "es") continue;
      if (term === candidate + "s" || term === candidate + "es") continue;

      const dist = boundedLevenshtein(candidate, term, 2);
      if (dist === 0 || dist > 2) continue;

      if (term.length <= 5) {
        if (dist > 1) continue;
        if (candidate[0].toLowerCase() !== term[0].toLowerCase()) continue;
      }

      if (!candidates_by_term[term]) candidates_by_term[term] = [];
      if (!candidates_by_term[term].includes(candidate))
        candidates_by_term[term].push(candidate);
    }
  }

  for (const [term, candidates] of Object.entries(candidates_by_term)) {
    candidates.sort();
    const c_str = candidates.map((c) => `'${c}'`).join(", ");
    diagnostics.push(`[Info] Possible Typos for '${term}': Found ${c_str}`);
  }

  function diag_sort_key(msg: string) {
    if (msg.startsWith("[Error]")) return 0;
    if (msg.startsWith("[Warning]")) return 1;
    return 2;
  }

  diagnostics.sort((a, b) => {
    const keyA = diag_sort_key(a);
    const keyB = diag_sort_key(b);
    if (keyA !== keyB) return keyA - keyB;
    return a.localeCompare(b);
  });

  return [definitions, diagnostics, raw_anchors];
}

/**
 * Inspects word/settings.xml for privacy flags that cause Microsoft Word to
 * silently strip authorship metadata (w:author, w:date, w:initials) from
 * tracked changes and comments on the next save. These flags do not affect
 * what this engine writes — but they will scrub the engine's attribution the
 * next time Word opens and saves the file, breaking auditability and
 * multi-turn agent state tracking.
 *
 * Returns a list of human-readable warning lines (without leading bullets);
 * empty when no relevant flag is present or all flags are explicitly disabled.
 */
export function extract_document_settings_warnings(
  doc: DocumentObject,
): string[] {
  const warnings: string[] = [];
  const settingsPart = doc.pkg.getPartByPath("word/settings.xml");
  if (!settingsPart) return warnings;

  const isTruthy = (el: Element): boolean => {
    // OOXML boolean rule: element present with no w:val attribute defaults to true.
    // w:val of "0", "false", or "off" means disabled. Anything else (including
    // "1", "true", "on") is enabled.
    if (!el.hasAttribute("w:val")) return true;
    const val = (el.getAttribute("w:val") || "").toLowerCase();
    return val !== "0" && val !== "false" && val !== "off";
  };

  const removePersonal = findDescendantsByLocalName(
    settingsPart._element,
    "removePersonalInformation",
  );
  if (removePersonal.length > 0 && isTruthy(removePersonal[0])) {
    warnings.push(
      "[Warning] Privacy flag `removePersonalInformation` is enabled in word/settings.xml. " +
        "Microsoft Word will strip the `w:author`, `w:initials`, and `w:date` attributes from every tracked change and comment the next time this document is opened and saved. " +
        "Edits made by this agent will lose attribution, breaking audit trails and any multi-turn workflow that relies on identifying prior edits.",
    );
  }

  const removeDateTime = findDescendantsByLocalName(
    settingsPart._element,
    "removeDateAndTime",
  );
  if (removeDateTime.length > 0 && isTruthy(removeDateTime[0])) {
    warnings.push(
      "[Warning] Privacy flag `removeDateAndTime` is enabled in word/settings.xml. " +
        "Microsoft Word will strip the `w:date` attribute from every tracked change and comment the next time this document is opened and saved. " +
        "Timestamps on this agent's edits will be lost on the next Word save.",
    );
  }

  return warnings;
}

/**
 * The appendix's `## Content Controls` block, or [] when unwarranted.
 *
 * Defensive: the appendix is advisory, so a malformed settings part or an
 * exotic control must not take down every read of the document.
 */
function content_controls_appendix_section(
  doc: DocumentObject,
  base_text: string,
): string[] {
  try {
    // Counts only. The appendix renders four numbers, and computing the full
    // ledger to get them added 115ms of unrendered value previews and
    // breadcrumbs to every read of a control-heavy document.
    const counts = fieldSummary(doc);
    const protection = readDocumentProtection(doc);
    return renderAppendixSection(
      counts,
      protection,
      'Read with mode="fields" for the full field ledger.',
    );
  } catch {
    return [];
  }
}

export function build_structural_appendix(
  doc: DocumentObject,
  base_text: string,
): string {
  const [defs, diagnostics, anchors] = extract_all_domain_metadata(
    doc,
    base_text,
  );
  const settings_warnings = extract_document_settings_warnings(doc);

  const lines: string[] = [
    "\n\n---",
    "",
    "<!-- READONLY_BOUNDARY_START -->",
    "# Document Structure (Read-Only)",
    "The content below is metadata describing the document's reference structure. Do not include this section in any tracked changes or edits \u2014 it is for your context only and will be discarded on write.",
  ];

  let has_content = false;

  if (settings_warnings.length > 0) {
    has_content = true;
    lines.push("\n## Document Settings");
    for (const warning of settings_warnings) {
      lines.push(`- ${warning}`);
    }
  }

  // Spec-fields-ledger §5: the HEADER LINES ONLY. The full ledger never
  // renders here — FedRAMP rev4 has 5,007 controls and the appendix is
  // bounded, so a ledger would swallow every other section.
  const cc_lines = content_controls_appendix_section(doc, base_text);
  if (cc_lines.length > 0) {
    has_content = true;
    lines.push("");
    lines.push(...cc_lines);
  }

  if (Object.keys(defs).length > 0) {
    has_content = true;
    lines.push("\n## Defined Terms");
    for (const [term, data] of Object.entries(defs)) {
      lines.push(
        `- "${term}" \u2014 used ${data.count} time${data.count === 1 ? "" : "s"}.`,
      );
    }
  }

  if (diagnostics.length > 0) {
    has_content = true;
    lines.push("\n## Semantic Diagnostics");
    for (const diag of diagnostics) {
      lines.push(`- ${diag}`);
    }
  }

  if (Object.keys(anchors).length > 0) {
    has_content = true;
    lines.push("\n## Named Anchors");
    for (const [b_name, data] of Object.entries(anchors)) {
      lines.push(`- ${b_name} \u2192 Anchored to: "${data.anchored_to}"`);
      for (const ref of data.referenced_from) {
        lines.push(`  - Referenced from: "${ref}"`);
      }
    }
  }

  if (has_content) {
    return lines.join("\n");
  }
  return "";
}
